### ISSUE TITLE
Moved provider from custom to Hashicorp registry. Plugin cache content causes death with invalid error 

### ISSUE THREAD

**Issue by Zordrak** [hashicorp/terraform#25763]
### Terraform Version
l

```
0.13.0-rc1
```

Many apologies, I don't have the debug output and I don't have time to stop and reproduce. If I find some time somewhere I will come back to it, but this ought to be easy to reproduce and the bug is ugly.

Had a custom provider on disk, forcefully invoked via .terraformrc with the versions.tf file telling it to pick it up.

That provider is now available in the registry so I did a replace-provider to move to the registry version, and then deleted the .terraformrc and the plugin. However, there was still a copy of the old provider in the old custom registry location in my plugin-cache (a different location, overridden by environment variable).

This completely broke terraform. It decided none of the existing providers in the plugin-cache could be used and redownloaded them all, and then after it redownloaded them, for each one it said (example):

```
Error: Failed to install provider
Error while installing hashicorp/aws v2.70.0: after installing
registry.terraform.io/hashicorp/aws it is still not detected in the target
directory; this is a bug in Terraform
```

Having deleted the old, should-no-longer-have-even-been-referenced provider from the plugin-cache everything was fine.

Tried to quickly reproduce the bug, not having the exact files or in the same place because I'm not sure where everything was, I just put a copy of the custom provider in the custom registry directory in the plugin-cache and it gives me:

```
Error: Unrecognized remote plugin message: 

This usually means that the plugin is either invalid or simply
needs to be recompiled to support the latest protocol.
```

Seems like when the plugin-cache is being walked, if it doesnt contain perfectly-as-expected content, even in the path for registries not being checked by the active code, things may explode?

**Comment by Zordrak** [comment]
After repeating the last plan with no change but turning off TF_LOG=TRACE, it worked fine.. I really don't know.

**Comment by danieldreier** [comment]
Hi! Thanks for reporting this. Based on the `this is a bug in Terraform` message, I think this is probably a valid issue, and I'd like to reproduce it locally. To do that, I have to be able to run this and run it on my workstation without inventing any details in order to be confident we're seeing the same behavior. As-is, there aren't really clear reproduction steps, and it sounds like you're also not able to reproduce it, so it isn't clear to me how an engineer could productively work on fixing it.

What I'd really need is a very concrete step-by-step reproduction case, with example `.terraformrc` and `versions.tf` files, the path on disk and environment variables you're setting, etc. My experience has been that when I try to take a stab at finding the exact reproduction case for issues like this, it can take a whole day and I might not stumble across the exact set of reproduction conditions - for example, I've worked on a similar issue where I didn't realize the reporter was running on Windows, and that the issue was related to incorrect handling of relative paths on Windows.

I'm happy to iterate on something together, but we need to start with a very concrete procedural list of specifics and build it up from there.

**Comment by mildwonkey** [comment]
hi! This is expected (but confusing) behavior - terraform expects to to manage the provider installation entirely, and behavior is undefined when the directory has been modified manually, as in this case. As you've already found, the solution is to remove the directory and allow terraform to re-populate it. 

We already have a feature request open to make the error message clearer, so I am going to close this as a duplicate. Thanks for reporting this, and sorry for the confusion!

**Comment by shamil** [comment]
It happens to me each time I try to upgrade provider(s). I get exact same error, and only works if I delete `.terraform` directory. I never touch this directory or alter it in any way. For me it's not expected behavior like @mildwonkey explains, it's the opposite it's very inconsistent behavior and broke many automated flows since terraform `0.13` upgrade.

**Comment by shamil** [comment]
I think it might be related to plugins cache dir (e,g `~/.terrafromd.d/plugins`)
As an example, here is how the `.terraform` directory looks like before I upgrade provider(s). **Pay attention to symlink**

```
.terraform/
└── plugins
    ├── registry.terraform.io
    │   └── hashicorp
    │       └── aws
    │           └── 3.1.0
    │               └── linux_amd64 -> /home/shamil/.terraform.d/plugins/registry.terraform.io/hashicorp/aws/3.1.0/linux_amd64
    └── selections.json
```

When I try to upgrade `aws` provider for example, I get this error:

```
terraform init

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "3.4.0"...

Error: Failed to query available provider packages

Could not retrieve the list of available versions for provider hashicorp/aws:
no available releases match the given constraints 3.4.0
```

Then I delete the `/home/shamil/.terraform.d` directory and to `terraform init` again and get:

```

Initializing the backend...

Initializing provider plugins...
- Finding hashicorp/aws versions matching "3.4.0"...
- Installing hashicorp/aws v3.4.0...

Error: Failed to install provider

Error while installing hashicorp/aws v3.4.0: after installing
registry.terraform.io/hashicorp/aws it is still not detected in the target
directory; this is a bug in Terraform
```

and the directory looks like below: 

Note that symlink is broken now, since I deleted the `/home/shamil/.terraform.d/` directory.

```
.terraform/
└── plugins
    ├── registry.terraform.io
    │   └── hashicorp
    │       └── aws
    │           ├── 3.1.0
    │           │   └── linux_amd64 -> /home/shamil/.terraform.d/plugins/registry.terraform.io/hashicorp/aws/3.1.0/linux_amd64
    │           └── 3.4.0
    │               └── linux_amd64
    │                   └── terraform-provider-aws_v3.4.0_x5
    └── selections.json
```

Something really confusing going on here, and if I delete both the `/home/shamil/.terraform.d` and the `.terraform` directory it starts working.

Prior terraform `0.13` this was working without the above mentioned issues.

**Comment by sajid-khan-js** [comment]
Looks like this fix in `0.13.1` might correct this: https://github.com/hashicorp/terraform/pull/25692

**Comment by shamil** [comment]
I was on `0.13.1` when posted and same behaviour was in `0.13.0`
